### ISSUE TITLE
Bug in proposer reward

### ISSUE THREAD

**Issue by djrtwo** [ethereum/eth2.0-specs#906]
When applying proposer rewards within an epoch transition in `get_justification_and_finalization_deltas`, we need to get the proposer index from the _previous_ epoch. Right now there is an assertion in `get_beacon_proposer_index` that it has to be from the current epoch. This was added when we changed the sampling of proposer to be based upon balance because the balance can only be deterministically know in the current epoch.

If we remove the assertion, we can't guarantee that the proposer we fetch from the previous epoch is actually correct (although it will be 99.9% of the time).

```python
# Proposer bonus
if index in get_attesting_indices(state, state.previous_epoch_attestations):
    proposer_index = get_beacon_proposer_index(state, inclusion_slot(state, index))
```

```python
def get_beacon_proposer_index(state: BeaconState,
                              slot: Slot) -> ValidatorIndex:
    current_epoch = get_current_epoch(state)
    assert slot_to_epoch(slot) == current_epoch
```

Going to bed. Will think about this tomorrow

**Comment by djrtwo** [comment]
@JustinDrake Suggests just giving proposer rewards immediately. We already do this with Transfers and with Slashing rewards. This avoids the issues of having to calculate proposers outside of the epoch in which they propose and adds negligible effects to state hashing.

cc: @vbuterin

**Comment by JustinDrake** [comment]
So @djrtwo found out that we must process proposer rewards immediately because every block can change the balance.

**Comment by JustinDrake** [comment]
Fixed in #920.
